### PR TITLE
Add random peer selector

### PR DESCRIPTION
**Pull request by kriskowal**
This change finishes the collection of basic peer selection strategies for
YARPC.

- random
- round-robin
- fewest-pending-requests

### CODE REVIEW

File: peer/randpeer/random.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	return &randomList{
+		subscribers: make([]*subscriber, 0, cap),
+		random:      rand.New(source),
+	}
+}
+
+var _ peerlist.Implementation = (*randomList)(nil)
+
+func (r *randomList) Add(peer peer.StatusPeer, _ peer.Identifier) peer.Subscriber {
+	index := len(r.subscribers)
```
[the remark this answers is not in the document]
Comment (kriskowal) [reply peer/randpeer/random.go]: The peerlist calls these methods under locks. The Implementation can be lock free.

File: peer/randpeer/list.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+func Capacity(capacity int) ListOption {
+	return listOptionFunc(func(options *listOptions) {
+		options.capacity = capacity
+	})
+}
+
+// Seed specifies the seed for generating random choices.
+func Seed(seed int64) ListOption {
+	return listOptionFunc(func(options *listOptions) {
+		fmt.Printf("HERE %d\n", seed)
```
Comment (AllenLuUber) [inline peer/randpeer/list.go]: nit: remove

File: peer/randpeer/random.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+type subscriber struct {
+	index int
+	peer  peer.StatusPeer
+}
+
+var _ peer.Subscriber = (*subscriber)(nil)
+
+func (s *subscriber) NotifyStatusChanged(peer.Identifier) {}
```
Comment (AllenLuUber) [inline peer/randpeer/random.go]: nit: `func (*subscriber) NotifyStatusChanged(peer.Identifier) {}`

File: peer/randpeer/list.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// Source is a source of randomness for the peer list.
+func Source(source rand.Source) ListOption {
+	return listOptionFunc(func(options *listOptions) {
+		options.source = source
+	})
+}
+
+// New creates a new round robin peer list.
```
Comment (AllenLuUber) [inline peer/randpeer/list.go]: nit: new random peer list

File: peer/randpeer/list.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+			newRandomList(options.capacity, options.source),
+			plOpts...,
+		),
+	}
+}
+
+// List is a PeerList that rotates which peers are to be selected randomly
+type List struct {
+	*peerlist.List
+}
```
Comment (AllenLuUber) [inline peer/randpeer/list.go]: why we do we need this struct? What is the problem of returning `*peerlist.List` in the `New`?

File: peer/randpeer/list.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+// data structures for this list.
+//
+// Defaults to 10.
+func Capacity(capacity int) ListOption {
+	return listOptionFunc(func(options *listOptions) {
+		options.capacity = capacity
+	})
+}
+
+// Seed specifies the seed for generating random choices.
```
Comment (AlexAPB) [inline peer/randpeer/list.go]: maybe comment that we default to `time.Now().UnixNano()` as the seed if this option is unspecified.

File: peer/randpeer/list.go (hunk 2 above)
Comment (kriskowal) [reply peer/randpeer/list.go]: ✔

File: peer/randpeer/list.go (hunk 6 above)
Comment (kriskowal) [reply peer/randpeer/list.go]: ✔

File: peer/randpeer/list.go (hunk 2 above)
Comment (kriskowal) [reply peer/randpeer/list.go]: ✔

File: peer/randpeer/list.go (hunk 4 above)
Comment (kriskowal) [reply peer/randpeer/list.go]: ✔

File: peer/randpeer/list.go (hunk 5 above)
Comment (kriskowal) [reply peer/randpeer/list.go]: This is necessary to "inherit" the methods of the embedded type.

File: peer/randpeer/random.go (hunk 3 above)
Comment (kriskowal) [reply peer/randpeer/random.go]: ✔

File: peer/randpeer/random.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	index := len(r.subscribers)
+	r.subscribers = append(r.subscribers, &subscriber{
+		index: index,
+		peer:  peer,
+	})
+	return r.subscribers[index]
+}
+
+func (r *randomList) Remove(peer peer.StatusPeer, _ peer.Identifier, sub peer.Subscriber) {
+	if sub, ok := sub.(*subscriber); ok && len(r.subscribers) > 0 {
```
Comment (abhinav) [inline peer/randpeer/random.go]: optional nit: invert if to reduce nesting

**Pull request merged**